app/vmalert/main.go

```diff
@@ -44,15 +63,69 @@ func main() {
 	if err := httpserver.Stop(*httpListenAddr); err != nil {
 		logger.Fatalf("cannot stop the webservice: %s", err)
 	}
+	cancel()
 	w.stop()
 }
 
 type watchdog struct {
-	storage *datasource.VMStorage
+	stg *datasource.VMStorage
+	p   provider.AlertProvider
+}
+
+func (w *watchdog) run(ctx context.Context, a config.Group, evaluationInterval time.Duration) {
+	t := time.NewTicker(evaluationInterval)
+	var metrics []datasource.Metric
+	var err error
+	var alerts []provider.Alert
+	defer t.Stop()
+	for {
+		select {
+		case <-t.C:
+			for _, r := range a.Rules {
+				if metrics, err = w.stg.Query(ctx, r.Expr); err != nil {
+					logger.Errorf("error reading metrics %s", err)
+					continue
+				}
+				// todo check for and calculate alert states
+				if len(metrics) > 0 {
+					alerts = provider.AlertsFromMetrics(metrics, a.Name, r)
```

Comment:
do we check for `for` parameter?

Reply:
not yet, going to add this check next PR